### DISCUSSION TITLE
Contributing my german voice for tts

### DISCUSSION THREAD

**Discussion by JRMeyer** [coqui-ai/TTS#114] [Archived]
<USER>>>> mrthorstenm<br/> <DATE> [November 6, 2019, 6:00am]<br/> <br/> Hello.<br/> <br/> My name is Thorsten Müller, native german speaker and i currently use<br/> mimic-recording-studio for recording my voice for tts generation. *slash* <br/> I'm using a corpus created by mycroft community member (gras64) taken<br/> phrases from<br/> <https://raw.githubusercontent.com/mozilla/voice-web/master/server/data/de/sentence-collector.txt><br/> and have recorded 7k phrases (from 30k) with a duration of round about 6<br/> hours at the moment. *slash* <br/> I want to contribute these ljspeech data (metadata.csv and wav files) to<br/> the community.<br/> <br/> Information and download on:<br/> <https://github.com/thorstenMueller/deep-learning-german-tts><br/> <br/> Hopefully it's useful for somebody.<br/> <br/> Thorsten<br/> <br/> [](https://discourse.mozilla.org/t/are-there-any-projects-dealing-with-artificial-speech/62352/2)<br/> <br/> [Are there any projects dealing with artificial speech<br/> <br/> [](https://discourse.mozilla.org/t/testfigures-doesnt-align/56807/22)<br/> <br/> [TestFigures doesn *slash* 't align<br/> <br/> [This is an archived TTS discussion thread from discourse.mozilla.org/t/contributing-my-german-voice-for-tts]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [November 6, 2019, 12:54pm]<br/> <br/> That's a great contribution thx. I'll share some results and feedback<br/> asap.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> mrthorstenm<br/> <DATE> [November 6, 2019, 1:04pm]<br/> <br/> You're welcome. Thanks for planning to share results when tested with<br/> it. *slash* <br/> I'm still recording at the moment and will update my wav files on google<br/> drive when i reached 10k recordings.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> mrthorstenm<br/> <DATE> [January 1, 2020, 10:30pm]<br/> <br/> Happy new year dear community<br/> <br/> Since i'm still recording my voice for community contribution (for<br/> several month now) i want to give a short update. I've recorded 12600<br/> phrases with a total audio length of 11 hours.<br/> <br/> ![](https://github.githubassets.com/favicon.ico){.site-icon width='32'<br/> [GitHub](https://github.com/thorstenMueller/deep-learning-german-tts)<br/> <br/> ### [thorstenMueller/deep-learning-german-tts](https://github.com/thorstenMueller/deep-learning-german-tts)<br/> <br/> Training files (csv and wav audio) for deep learning of tts/stt in<br/> german language. - thorstenMueller/deep-learning-german-tts<br/> <br/> *slash* <br/> Direct dataset download:<br/> <https://drive.google.com/open?id=1NTi-4r3EWl5dw0k2o4Xh92G0OHvhoxAJ><br/> <br/> Results of analyze.py:<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> mrthorstenm<br/> <DATE> [January 8, 2020, 6:52pm]<br/> <br/> After performing a training with 100k steps and 14306 recorded phrases I<br/> found that the quality was not as desired. Dominik<br/> ) and Eltonico from the<br/> Mycroft Forum were kind enough to check the quality of my recordings. It<br/> turned out that some recordings had reverberation and echoes and<br/> therefore not ideal for TTS training. *slash* <br/> Together with Dominik I try to identify and optimize the bad files. When<br/> that is complete, I will provide the link to the cleaned and optimized<br/> dataset here. *slash* <br/> Many thanks to Dominik and Eltocino for their support in this matter.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> mrthorstenm<br/> <DATE> [January 10, 2020, 7:46pm]<br/> <br/> he<br/> recommended to document the progress and lessons learned for the<br/> community. As i think this is a good idea i (and probable Dominik) will<br/> update this thread on a regular basis.<br/> <br/> At the moment i listen to all my recordings and categorize all wavs in<br/> green (good), yellow (needs revision), red (removed from dataset) while<br/> Dominik starts optimizing the files.<br/> <br/> 20.34.28.filename}[2134×734 187<br/> <br/> After removing the red ones, and doing some optimization by Dominik on<br/> the yellow ones we hopefully have an acceptable dataset for german tts<br/> generation.<br/> <br/> I uploaded some recorded samples on my github page<br/> (<https://github.com/thorstenMueller/deep-learning-german-tts>) just for<br/> interested people to get an impression of the sound of my voice.<br/> <br/> Lessons learned so far:<br/> <br/> (little-endian) data, WAVE audio, Microsoft PCM, 16 bit, stereo<br/> 44100 Hz' which is higher sample rate than required (16000-22050Hz).<br/> Stereo should not be needed.<br/> mentioned in several places, so please take it serious)<br/> <br/> Even if i use this thread for documenting the progress this should not<br/> be a soliloquy thread so feedback of every kind is welcome.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> nmstoker<br/> <DATE> [January 10, 2020, 10:22pm]<br/> <br/> - I look<br/> forward to hearing more on the techniques.<br/> <br/> You might also be interested in the speaker embedding code in the dev<br/> branch here: <https://github.com/mozilla/TTS/tree/dev/speaker_encoder><br/> <br/> Using it with my own single speaker dataset I was able to identify a<br/> small but distinct cluster of my recordings which were more muffled than<br/> the majority. Curiously it was picking up that I had a mild cold with<br/> some of the audio (which at the time I'd thought wasn't audible but<br/> turned out to be) That was impacting the quality of the output audio<br/> from my trained model and I saw an improvement when I pruned it.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> mrthorstenm<br/> <DATE> [January 11, 2020, 10:28am]<br/> <br/> for the compliment<br/> and the link to speaker_encoder. *slash* <br/> I'm gonna give it a try as soon as i removed the obviously bad files<br/> from the dataset.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> mrthorstenm<br/> <DATE> [January 27, 2020, 10:34pm]<br/> <br/> Even if there's nothing new to show right now (no new dataset upload<br/> yet), i'm not lazy on this topic and want to keep this thread updated.<br/> <br/> 1. I removed really bad files from the dataset<br/> 2. I converted the files to mono and 22k sample rate (from stereo and<br/> 44k sample rate)<br/> listened to the 'yellow'<br/> files and is optimistic that he can optimize most files (reverb and<br/> random noise) whenever his rare spare time allows to invest time on<br/> it<br/> 4. I borrowed/bought semi-professional or at least better microphone<br/> equipment and set up a better location for futher recordings.<br/> <br/> I continue on recording new phrases with a more optimized setup. So i<br/> and i) can<br/> present a new cleaner dataset in the future.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> mrthorstenm<br/> <DATE> [February 8, 2020, 12:01pm]<br/> <br/> Currently i've recorded further 2.000 phrases with new/better equipment<br/> and room situation (still recording) and after 1.300 phrases i started a<br/> 'quick' training run up to 72k steps. *slash* <br/> The result sounds not too bad, even it's still robotic. Due upload<br/> restrictions here i uploaded the sample wav file to my github account at<br/> <https://github.com/thorstenMueller/deep-learning-german-tts/blob/master/samples/tts_tacotron1/>.<br/> <br/> But what's wrong with my dataset because of the horizontal line at 60? *slash* <br/> <br/> I wanted to check the quality of my dataset with tool by<br/> on<br/> <https://github.com/mozilla/TTS/tree/dev/speaker_encoder>, but i<br/> struggle with this information:<br/> <br/> shuf metadata.csv > metadata_shuf.csv<br/> head -n 12000 metadata_shuf.csv > metadata_train.csv<br/> tail -n 1100 metadata_shuf.csv > metadata_val.csv<br/> <br/> Do these values (12000 and 1100) come from this dataset<br/> <https://keithito.com/LJ-Speech-Dataset/> which consists off 13.100<br/> audio clips. *slash* <br/> So would these values on a 10.000 clips ljspeech dataset mean 8.900<br/> (train) and 1.100 (validate) or on which calculation depend these<br/> values?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> nmstoker<br/> <DATE> [February 8, 2020, 6:31pm]<br/> <br/> . That section<br/> where you've got the shuf, head and tail commands is for setting up the<br/> LJ Speech dataset for training with the main part of the repo. That<br/> shouldn't be needed for using the Speaker Encoder notebook (although you<br/> certainly could use that dataset and look for similar audio samples).<br/> <br/> You're right that the 12,000 and 1,100 values come from the total size<br/> of LJ Speech, as those steps are effectively splitting the whole dataset<br/> into a training and a validation set.<br/> <br/> But if you simply want to look through your own audio, eg to find any<br/> unusual clusters, then you don't need to split it up, just work on all<br/> the audio together. You can just feed it all the embeddings from your<br/> entire dataset and then the notebook will plot them in Bokeh.<br/> <br/> To get the embeddings, have a look at compute_embeddings.py (reading it<br/> again now, this could be a little clearer in the instructions<br/> You basically point it at your audio files and it creates a whole load<br/> of .npy files that are then used by the notebook.<br/> <br/> I'm away from my computer right now, but let me know how you get on and<br/> if you're stuck I'm happy to help try to figure it out.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> mrthorstenm<br/> <DATE> [February 9, 2020, 7:17am]<br/> <br/> Thanks for your clarification<br/> <br/> Since i used the mycroft mimic2 train.py i've got no<br/> `best_model.pth.tar`. When i tried running train.py from the mozilla tts<br/> 'speaker_encoder' folder i ran into problems with umap. *slash* <br/> Removing umap and installing umap-learn 0.3.10 fixed this issue for me.<br/> <br/> So i'm currently running a training with the mozilla version of<br/> train.py.<br/> <br/> Am i right that a training-run (until which step is wise?) is required<br/> for computer_embeddings.py?<br/> <br/> Currently i get the following error when try to run:<br/> <br/> python3 ./compute_embeddings.py model_path /home/thorsten/___dev/libri_tts/speaker_encoder/libritts_360-half-February-08-2020_10+27PM-e37503c/best_model.pth.tar config_path config.json<br/> <br/> Traceback (most recent call last):<br/> File './compute_embeddings.py', line 41, in <module><br/> c = load_config(args.config_path)<br/> File '/home/thorsten/___dev/mozilla/TTS/tts_namespace/TTS/utils/generic_utils.py', line 23, in load_config<br/> input_str = f.read()<br/> File '/usr/lib/python3.6/codecs.py', line 321, in decode<br/> (result, consumed) = self._buffer_decode(data, self.errors, final)<br/> UnicodeDecodeError: 'utf-8' codec can't decode byte 0x80 in position 0: invalid start byte<br/> <br/> But i'm gonna check this issue tomorrow. *slash* <br/> Thank you and good night for now<br/> <br/> edit:<br/> <br/> stays.<br/> <br/> }<br/> <br/> ```<br/> python3 ./compute_embeddings.py model_path /home/thorsten/___dev/libri_tts/speaker_encoder/libritts_360-half-February-08-2020_10+27PM-e37503c/best_model.pth.tar config_path ./configUNKNOWN.json<br/> Traceback (most recent call last):<br/> File './compute_embeddings.py', line 41, in <module><br/> c = load_config(args.config_path)<br/> File '/home/thorsten/___dev/mozilla/TTS/tts_namespace/TTS/utils/generic_utils.py', line 23, in load_config<br/> input_str = f.read()<br/> File '/usr/lib/python3.6/codecs.py', line 321, in decode<br/> (result, consumed) = self._buffer_decode(data, self.errors, final)<br/> UnicodeDecodeError: 'utf-8' codec can't decode byte 0x80 in position 0: invalid start byte<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> nmstoker<br/> <DATE> [February 9, 2020, 11:38pm]<br/> <br/> Edit: sorry, I don't know how but earlier I think I misunderstood what<br/> you were doing (so what I wrote before probably wasn't helpful<br/> sorry!)<br/> <br/> It might be easiest to have a go using it with the libritts 360 or 100<br/> files (available here: <http://www.openslr.org/60/> **NB:** 360 set is<br/> 27Gb!) just so you can be sure you've recreated what's in the notebook<br/> and *only then swap in your files*.<br/> <br/> I don't think you'd need to do the training step - you can try the<br/> pretrained model 'as is' first with computer_embeddings.py first just to<br/> confirm that you can get the embeddings and then use the notebook.<br/> Roughly it should be something like this:<br/> <br/> 1. Start with the pretrained model<br/> 2. Use it to create the embeddings for LibriTTS files<br/> 3. Plot these using the notebook (<br/> [PlotUmapLibriTTS.ipynb](https://github.com/mozilla/TTS/blob/master/speaker_encoder/notebooks/PlotUmapLibriTTS.ipynb))<br/> 4. Go back to step 2 to create embeddings for your audio and try step 3<br/> 5. Explore to see that audio is grouped by similar recordings - in my<br/> case I had one main cluster, with the next biggest cluster being an<br/> offshoot that had muffled audio.<br/> 6. If there aren't any discernible patterns to your clusters with your<br/> audio, then it is possible that it would need to be re-trained (but<br/> that wasn't something I needed to do myself)<br/> <br/> BTW: you are correct to install umap-learn<br/> <br/> Not sure about the utf-8 errors<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> nmstoker<br/> <DATE> [February 9, 2020, 10:15pm]<br/> <br/> It might also be worth having a look at<br/> <https://github.com/resemble-ai/Resemblyzer> too as that enables similar<br/> clustering.<br/> <br/> What this version in TTS does is use the pretrained model Eren created<br/> along with the interactive plotting where you can listen to the audio by<br/> clicking on the chart (thus easily attempt to discover what connects the<br/> clusters)<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> mrthorstenm<br/> <DATE> [February 10, 2020, 6:25am]<br/> <br/> Thanks for your responses.<br/> <br/> The confusion is my fault since my code snipplets shows the string<br/> 'libritts_360'. But that's just the name in the default config.yaml. *slash* <br/> I use an **ljspeech** dataset. But you're right. I will check my setup<br/> with a pretrained model first just to be sure all libs and dependencies<br/> are correct before trying to verify my own ljspeech dataset.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> smg<br/> <DATE> [February 11, 2020, 6:08pm]<br/> <br/> > omputer_embeddings.py<br/> <br/> Dear Neil, *slash* <br/> I wonder if you could be as helpful for italian language as you have<br/> been here for german. *slash* <br/> I am trying to replicate steps suggested, *slash* <br/> so I downloaded LibriTTS 100 version (english) - OK<br/> <br/> 1. Start with the pretrained model<br/> 2. Use it to create the embeddings for LibriTTS files *slash* <br/> Here I start having some problems uderstanding your suggestion. *slash* <br/> I have read in Multi Speaker Embeddings<br/> <https://github.com/mozilla/TTS/issues/166> *slash* <br/> then I went to<br/> <https://github.com/mozilla/TTS/tree/dev/speaker_encoder> *slash* <br/> and tried to follow instructions there like:<br/> <br/> Download a pretrained model from [Released<br/> Models](https://github.com/mozilla/TTS/wiki/Released-Models) page.<br/> <br/> To run the code, you need to follow the same flow as in TTS.<br/> <br/> should match your TTS model.<br/> `python speaker_encoder/train.py --config_path speaker_encoder/config.json --data_path ~/Data/Libri-TTS/train-clean-360`<br/> `python speaker_encoder/compute_embeddings.py --use_cuda true /model/path/best_model.pth.tar model/config/path/config.json dataset/path/ output_path`<br/> . This code parses all .wav files at the given dataset path and<br/> generates the same folder structure under the output path with the<br/> generated embedding files.<br/> <br/> As step 1 and 2 are the start you suggest, could you give some more<br/> detailed guidance so that I can proceed?<br/> <br/> thanks<br/> <br/> Sergio<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> smg<br/> <DATE> [February 11, 2020, 6:51pm]<br/> <br/> Hi, *slash* <br/> adding on my previous post, let we add this additional piece of usage<br/> help info coming from: *slash* <br/> <https://github.com/mozilla/TTS/wiki/Speaker-Encoder:><br/> <br/> TTS has a subproject, called [Speaker<br/> Encoder](https://github.com/mozilla/TTS/tree/dev/speaker_encoder). *slash* <br/> It is an implementation of <https://arxiv.org/abs/1710.10467> . There is<br/> also a released model trained on LibriTTS dataset with *slash* ~1000 speakers<br/> in [Released<br/> Models](https://github.com/mozilla/TTS/wiki/Released-Models) page.<br/> <br/> Ok, so I go to the Released-Models page and I try to choose the good<br/> one. *slash* <br/> But where or which one is the one you would suggest? *slash* <br/> I would go for the Tacotron-iter-170K defined as 'More stable and longer<br/> trained model.' *slash* <br/> But if I get it right this is for LJSpeech, a different set than the<br/> LibriTTS you suggest using whose 'corpus consists of 585 hours of speech<br/> data at 24kHz sampling rate from 2,456 speakers and the corresponding<br/> texts'.<br/> <br/> LJSpeech instead is a 'speech dataset consisting of 13,100 short audio<br/> clips of a single speaker reading passages from 7 non-fiction books. A<br/> transcription is provided for each clip. Clips vary in length from 1 to<br/> 10 seconds and have a total length of approximately 24 hours. The texts<br/> were published between 1884 and 1964, and are in the public domain. The<br/> audio was recorded in 2016-17 by the LibriVox project and is also in the<br/> public domain.'.<br/> <br/> So, at the end, can you help in clarifying this inconsistency? *slash* <br/> Or am I messing up things between<br/> [TTS](https://github.com/mozilla/TTS/tree/dev)/ **speaker_encoder**<br/> github instructions *slash* <br/> and and those in the wiki at<br/> <https://github.com/mozilla/TTS/wiki/Speaker-Encoder> *slash* <br/> being a newcomer here?<br/> <br/> thanks for your help *slash* <br/> sergio<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> sanjaesc<br/> <DATE> [February 12, 2020, 9:24am]<br/> <br/> > Ok, so I go to the Released-Models page and I try to choose the good<br/> > one. *slash* <br/> > But where or which one is the one you would suggest?<br/> <br/> There is a pre-trained Speaker Encoder Model in the middle.<br/> <br/> Link:<br/> <https://drive.google.com/open?id=1CwXxwm0qKk2lcPL_4EQDirD8XMyvHIqR><br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> smg<br/> <DATE> [February 12, 2020, 1:40pm]<br/> <br/> Thanks, sorry for the inconvenience. *slash* <br/> As posted elsewhere I am trying to setup a first italian language<br/> attempt starting from the voices collectd in the Mozilla Common Voice<br/> Project. *slash* <br/> As far as I read around here due to the amount of different voices and<br/> difference in quality, we should not have too high expectations, *slash* <br/> Then I would also start a single italian voice attempt. *slash* <br/> thanks again *slash* <br/> sergio<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> nmstoker<br/> <DATE> [February 12, 2020, 9:04pm]<br/> <br/> Hi [ *slash* smg](<br/> help but it would be good to confirm your objective with the speaker<br/> embeddings here as there are some different ways it can be used.<br/> <br/> You mention that you're trying to use the various Italian speakers from<br/> Common Voice. Are you trying to produce a) a multi speaker model as per<br/> issue [ *slash* #166] or b) trying to train a single TTS but simply<br/> using all the audio from a range of different speakers?<br/> <br/> If a): I haven't looked at multi speaker models myself so I would have<br/> to defer to the expertise of others, but the impression I got from<br/> reading [ *slash* #166] is that it's not able to produce such clear<br/> voices (yet)<br/> <br/> If b): That's not something I've tried, but with my recordings of a<br/> single speaker (me!) I've seen that including audio where I haven't<br/> spoken in a really consistent speaking style results in a big impact on<br/> output voice quality, so I'm thinking it would be unlikely to work well<br/> at all where the speakers were different people.<br/> <br/> The part I expect I can help with is if you're looking at using the<br/> speaker embeddings tools to help analyse the audio you're planning to<br/> use for training a TTS model.<br/> <br/> If that's what you want then I'd be trying to guide you to recreate<br/> what's shown in the video for the multi speaker audio or for a single<br/> speaker.<br/> <br/> To do that you'd need the pertained model from Released Models. It is<br/> showed.<br/> <br/> You'd want to go to the **/speaker_encoder** directory and adjust the<br/> paths in config.json so they pointed at your audio for the datasets path<br/> and at a suitable location for the output_path (this is where you'll be<br/> saving the embeddings, with are .npy files that correspond to each of<br/> your audio files). You can also just pass these paths as parameters to<br/> compute_embeddings.py too, so I'm not sure strictly how necessary it is<br/> to change them in config.json. Also to clarify, the contents of the<br/> config.json in that directory are currently the same as the config.json<br/> that comes with the model, but I expect you'd want to use the ones<br/> supplied with the model you're using if they were to differ (for<br/> instance if you'd trained a speaker embedding model with different<br/> parameters)<br/> <br/> Then run **compute_embeddings.py**, passing the relevant arguments (so<br/> the use_cuda parameter and then the model you downloaded above (as per<br/> ), the config and the<br/> paths. This works through your audio files and produces the .npy files I<br/> mentioned above.<br/> <br/> Then go into the Notebook directory and launch the Jupiter notebook.<br/> It'll need some basic edits to point the paths at your files, but this<br/> should be fairly obvious (eg MODEL_RUN_PATH and the others must point to<br/> somewhere on your computer). Depending on where you stored your<br/> embeddings you may need to edit that glob command for embed_files.<br/> <br/> Also under **Set up the embeddings**, comment/uncomment the relevant<br/> part for single speaker or multispeaker.<br/> <br/> Finally you should be in a position to step through the Notebook<br/> <br/> If all goes smoothly you should see the Bokeh chart near the end of the<br/> notebook. Don't forget to run the final cell, which starts a local<br/> server, so that the hyperlinks on each plotted embedding point to the<br/> corresponding audio file (this is the way that you can then click on the<br/> various sections of the chart and hear what the corresponding audio<br/> sounds like, as is shown in the video)<br/> <br/> So, that's quite a few steps! I hope I've got it all right and been<br/> clear. Have a go and if you run into any problems you can't figure out,<br/> report back as much useful detail as you can and I'll do my best to<br/> assist<br/> <br/> [**[next page<br/> →](https://discourse.mozilla.org/t/contributing-my-german-voice-for-tts/48150?page=2)**<br/> <br/> [Archived Post]<br/>

**Comment by thorstenMueller** [comment]
Hi.

Not sure if it's smart to reply to an archied thread from Mozilla discourse but that discussion/topic started there (https://discourse.mozilla.org/t/contributing-my-german-voice-for-tts/48150/225).

In addition to my open "Thorsten" german dataset i've recorded an emotional dataset based on 300 phrases recorded in following emotions.

* Normal :slightly_smiling_face:
* Disgusted :nauseated_face:
* Angry :angry:
* Amused :grinning:
* Surprised :astonished:
* Sleepy :pensive:

Currently it's audio optimization time (_thx @domcross_) but once it's ready i'll publish it.
Two samples (emotional comparison) can be found here https://github.com/thorstenMueller/deep-learning-german-tts/issues/13.

Please always keep in mind that i'm no professional voice artists.
